### ISSUE TITLE
Better terminal detection in default config

### ISSUE THREAD

**Issue by blueyed** [awesomeWM/awesome#1217]
One annoyance when being dropped into the default config is that you are getting xterm as your terminal.

E.g. with Debian based systems, we could default to `x-terminal-emulator`.

I am using urxvt myself, but there seems to be nothing like `x-terminal-emulator` for Arch Linux, or is there?

**Comment by actionless** [comment]
ugly but working for me:
```bash
awesome-client 'require("awful").spawn(\'bash -c "unknown_term || xterm || urxvt"\')'
```

**Comment by blueyed** [comment]
Where is `unknown_term` coming from?

**Comment by actionless** [comment]
it's an example :)

**Comment by actionless** [comment]
or we could try to do it cleaner but still weird, like:

```lua

local function get_term(callback)
  local found = false
  for _, term in ipairs({'xterm', 'urxvt', 'st', 'whatever else'}) do
    awful.spawn.easy_async('which ' .. term, function(_, _, _, exit_code)
      if exit_code == 0 then 
        found = true
      end
    end)
    if found then 
      callback(term)
      return
    end
  end
end

<...>
    awful.key({ modkey,           }, "Return", function () awful.spawn(terminal) end,
              {description = "open a terminal", group = "launcher"}),
```

**Comment by blueyed** [comment]
Yeah, I see - but in this example xterm would still  be preferred.
We should not over-engineer it,  but just try to look at what distributions provide.

**Comment by actionless** [comment]
ah, i've got the point

then we could allow it to override it somehow, see updated example

**Comment by actionless** [comment]
btw, do we still have some synchronous spawn? it could be easier to use `get_terminal` as synchronous function which returns terminal

**Comment by actionless** [comment]
regarding the latter point -- i think it would be better to use `file:///usr/share/doc/awesome/doc/index.html` instead

**Comment by funeralismatic** [comment]
>file:///usr/share/doc/awesome/doc/index.html instead

Only with docs installed, otherwise we need to go to wiki website. So we need yet another parser. Useless, don't need.

Link in man page should be enough.

**Comment by actionless** [comment]
built documentation will match the installed awesome version while online documentation will not

**Comment by psychon** [comment]
> btw, do we still have some synchronous spawn? it would be easier to use get_terminal() as synchronous function which returns terminal

Which synchronous functions were available? The only one that I remember is `awful.util.pread` and that was just some small wrapper around `io.popen`: https://github.com/awesomeWM/awesome/blob/061751dd9d3a3f08e4efa1739adb4f93e73bee98/lib/awful/util.lua.in#L93-L107

**Comment by actionless** [comment]
yup, that was the one i kept in mind

what do you think of using it for cached sycnhronous get_terminal() function? (see asynchronous example in one of my comments above: https://github.com/awesomeWM/awesome/issues/1217#issuecomment-259575577)

**Comment by psychon** [comment]
I would think that it would not help you / would not work.
You cannot get the exit code with `popen` (you can not even redirect stderr, so all unsuccessful lookups would spam awesome's stderr with something like "sh: sl: command not found").

**Comment by actionless** [comment]
@Elv13 mb glib have some analogue of `which` command?
